ambry-replication/src/main/java/com/github/ambry/replication/BackupCheckerThread.java

```diff
@@ -0,0 +1,165 @@
+ * its peers and this checker will detect such lagging replicas as well.
+ * TODO: Since we are just checking for existence of a blob in local-store or backup, we don't need to fetch the entire
+ * TODO: blob from on-prem server. Consider overriding that method and getting just a checksum from on-prem server.
+ */
+public class BackupCheckerThread extends ReplicaThread {
+
+  private final Logger logger = LoggerFactory.getLogger(BackupCheckerThread.class);
+
+  public BackupCheckerThread(String threadName, FindTokenHelper findTokenHelper, ClusterMap clusterMap,
+      AtomicInteger correlationIdGenerator, DataNodeId dataNodeId, ConnectionPool connectionPool,
+      ReplicationConfig replicationConfig, ReplicationMetrics replicationMetrics, NotificationSystem notification,
+      StoreKeyConverter storeKeyConverter, Transformer transformer, MetricRegistry metricRegistry,
+      boolean replicatingOverSsl, String datacenterName, ResponseHandler responseHandler, Time time,
+      ReplicaSyncUpManager replicaSyncUpManager, Predicate<MessageInfo> skipPredicate,
+      ReplicationManager.LeaderBasedReplicationAdmin leaderBasedReplicationAdmin) {
+    super(threadName, findTokenHelper, clusterMap, correlationIdGenerator, dataNodeId, connectionPool,
+        replicationConfig, replicationMetrics, notification, storeKeyConverter, transformer, metricRegistry,
+        replicatingOverSsl, datacenterName, responseHandler, time, replicaSyncUpManager, skipPredicate,
+        leaderBasedReplicationAdmin);
+  }
+
+  /**
+   *  This checks for GDPR compliance.
+   *  If a blob has been deleted from on-prem servers, it must be absent in backups.
+   *  If it is in TEMP or PERM state, then backup is behind servers & DELETE has not propagated to the backup.
+   * @param onPremBlob the {@link MessageInfo} that will be transformed into a delete-operation
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyDelete(MessageInfo onPremBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.noneOf(MessageInfoType.class);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.of(StoreErrorCodes.ID_Deleted,
+        StoreErrorCodes.ID_Deleted_Permanently, StoreErrorCodes.ID_Not_Found);
+    checkBackup(onPremBlob, remoteReplicaInfo, acceptableBackupBlobStates, acceptableStoreErrorCodes);
+  }
+
+  /**
+   * This method checks if permanent blob is permanent in backups as well.
+   * This method does not expect any errors when retrieving the blob from store.
+   * @param onPremBlob the {@link MessageInfo} that will be transformed into a TTL update
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyTtlUpdate(MessageInfo onPremBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.of(MessageInfoType.TTL_UPDATE);
```

Comment:
What's our design for the applyXXX?
We know whenever applyXXX is applied, the local store is out of sync. It's an issue already for the "sealed" partition.
We still set the acceptableBackupBlobStates and acceptableStoreErrorCodes and use them to compare. Is that redundant?

Reply:
You are right that when we come here the remote and local replica are out of sync. And for a sealed partition, the states will never be the same. But for an un-sealed partition, the remote replica we are comparing with backup may be behind the backup. For eg, a ttl-update may not have propagated to a replica but may have propagated to the backup. I want to relax the constraints of comparison at that point. 
Further, before reporting a missing blobId or a message, I want to be sure that its really missing. Hence, the additional checks. There is no impact on performance here.

Reply:
Further, for someone just reading this class, it may not be evident what we are doing if we just leave the applyXXX methods empty or just log an error since all the checks happen above it in some other class.